File: cache.py
import time
from threading import Timer
from constants import *

# The idea is to have a single Timer thread that will print when the oldest message has expired
#   Using a Timer thread since they can be cancelled, apparently threads can't be killed
# Use a queue for cache so you know the oldest message
# TODO: Need to add unique id to messages from the DeathStarBench message queue, and integrate with MessageCache
# TODO: Need to update the CacheEntry _extract_id method to work with JSON


# Manages the sleeper thread
# Start the sleeper on init
# Can cancel the sleeper
class WakeupThread:

    def __init__(self, sleepTime, entry) -> None:
        self.sleepTime = sleepTime
        self.entry = entry
        self.thread = self._start_timer()

    def _start_timer(self) -> Timer:
        wakeup_thread = Timer(
            self.sleepTime,
            self._thread_function_wrapper,
            args=[],
        )
        wakeup_thread.start()
        return wakeup_thread

    def _thread_function_wrapper(self):
        invalidation_thread(self.entry.get_id())
        self.entry.call_cache_timeout_handler()

    def cancel_timer(self) -> None:
        self.thread.cancel()


# Represents a single entry in the cache
# We only want the message's ID and the time it was sent
class CacheEntry:

    def __init__(self, id, message, cache_handler) -> None:
        self.id = id
        self.message = message
        self.cache_timeout_handler = cache_handler
        self.thread = WakeupThread(LIFETIME, self)

    def get_id(self) -> int:
        return self.id

    def call_cache_timeout_handler(self):
        return self.cache_timeout_handler()

    def kill_thread(self) -> None:
        self.thread.cancel_timer()


# Actual cache that holds the entries
class Cache:

    def __init__(self) -> None:
        self.collection: dict[int, CacheEntry] = {}

    # TODO: probably extracrt message['id'] once its JSON
    def _extract_id(self, message) -> int:
        return hash(message)

    def _delete(self, id) -> None:
        del self.collection[id]

    def add(self, message) -> None:
        id = self._extract_id(message)
        self.collection[id] = CacheEntry(id, message, lambda *_: self._delete(id))

    def remove(self, message) -> None:
        id = self._extract_id(message)
        self.collection[id].kill_thread()
        self._delete(id)


# This is just an interface for the message queue to use
class MessageCache:

    def __init__(self) -> None:
        self.cache = Cache()

    def add_sent_message(self, message) -> None:
        self.cache.add(message)

    def receive_message(self, message):
        self.cache.remove(message)

File: DeathStarBench/socialNetwork/src/cache.cpp
#include "cache.h"

int invalidation_count = 0;

void invalidation_thread(int id) {
	std::cout << "Message with ID: " << id << " has expired." << std::endl;
    invalidation_count++;
}

// WakeupThread Implementation
WakeupThread::WakeupThread(double sleepTime, CacheEntry* entry)
    : sleepTime(sleepTime), entry(entry), isCancelled(false) {
    thread = std::make_unique<std::thread>(&WakeupThread::threadFunction, this);
}

WakeupThread::~WakeupThread() {
    cancelTimer();
}

void WakeupThread::cancelTimer() {
    isCancelled = true;
    if (thread && thread->joinable()) {
        thread->detach();
    }
}

void WakeupThread::threadFunction() {
    std::this_thread::sleep_for(std::chrono::duration<double>(sleepTime));
    if (!isCancelled) {
        invalidation_thread(entry->getId());
        entry->callCacheTimeoutHandler();
    }
}

// CacheEntry Implementation
CacheEntry::CacheEntry(int id, std::string message, std::function<void()> cacheHandler)
    : id(id), message(std::move(message)), cacheTimeoutHandler(std::move(cacheHandler)) {
    thread = std::make_unique<WakeupThread>(LIFETIME, this);
}

int CacheEntry::getId() const {
    return id;
}

void CacheEntry::callCacheTimeoutHandler() {
    cacheTimeoutHandler();
}

void CacheEntry::killThread() {
    thread->cancelTimer();
}

// Cache Implementation
void Cache::add(Message message) {
    int id = extractId(message);
    collection[id] = std::make_unique<CacheEntry>(id, message.text, [this, id]() { this->removeById(id); });
}

void Cache::remove(Message message) {
    int id = extractId(message);
    if (collection.count(id)) {
        collection[id]->killThread();
        removeById(id);
    }
}

int Cache::extractId(const Message& message) {
    return message.id;
}

void Cache::removeById(int id) {
    collection.erase(id);
}

// MessageCache Implementation
void MessageCache::addSentMessage(Message message) {
    cache.add(message);
}

void MessageCache::receiveMessage(Message message) {
    cache.remove(message);
}
